infrastructure/src/test/java/org/corfudb/infrastructure/LogReplicationServerTest.java

```diff
@@ -126,19 +158,95 @@ public void testHandleLeadershipQuery() {
+        Assertions.assertThat(argument.getValue().getPayload().getLrLeadershipResponse().getIsLeader()).isFalse();
+
+        // verify the response when not the leader
+        replicationContext = new LogReplicationContext(mock(LogReplicationConfigManager.class),
+                0L, SAMPLE_HOSTNAME, false, mock(LogReplicationPluginConfig.class));
+        lrServer = spy(new LogReplicationServer(context, new HashSet<>(), topology, metadataManager,
+                replicationContext));
+        mockServerRouter = mock(IClientServerRouter.class);
+        lrServer.getHandlerMethods().handle(request, null, mockServerRouter);
+        argument = ArgumentCaptor.forClass(ResponseMsg.class);
+        verify(mockServerRouter, times(1)).sendResponse(argument.capture());
+        Assertions.assertThat(argument.getValue().getPayload().getLrLeadershipResponse().getIsLeader()).isFalse();
+
+        // verify the response when leader but the topology does not have FULL_TABLE
+        replicationContext = new LogReplicationContext(mock(LogReplicationConfigManager.class),
+                0L, SAMPLE_HOSTNAME, true, mock(LogReplicationPluginConfig.class));
+        when(topology.getRemoteSourceClusterToReplicationModels()).thenReturn(new HashMap<>());
+        lrServer = spy(new LogReplicationServer(context, new HashSet<>(), topology, metadataManager,
+                replicationContext));
+        mockServerRouter = mock(IClientServerRouter.class);
+        lrServer.getHandlerMethods().handle(request, null, mockServerRouter);
+        argument = ArgumentCaptor.forClass(ResponseMsg.class);
+        verify(mockServerRouter, times(1)).sendResponse(argument.capture());
+        Assertions.assertThat(argument.getValue().getPayload().getLrLeadershipResponse().getIsLeader()).isFalse();
+    }
+
+    /**
+     * Make sure that the LogReplicationServer will process {@link LogReplicationLeadershipRequestMsg}
+     * and provide an appropriate {@link ResponseMsg} message.
+     *
+     * Here both the source is on a older version, but the sink is in LR V2 version. Hence the request neither has the session information nor the source clusterID
+     * We test the following cases where the leadership request arrives:
+     * (i) at the leader and the sessionManager already knows about the session
+     * (ii) at the leader, but the sessionManager is yet to create the session.
+     * (iii) at leader node, but the topology does not contain a FULL_TABLE replication model
+     */
+    @Test
+    public void testHandleLeadershipQueryOnMixedVersion() {
+        final LogReplicationLeadershipRequestMsg leadershipQuery =
+                LogReplicationLeadershipRequestMsg.newBuilder().build();
+        final RequestMsg request =
+                getRequestMsg(HeaderMsg.newBuilder()
+                                .setVersion(getDefaultProtocolVersionMsg())
+                                .setIgnoreClusterId(true)
```

Comment:
I wonder why do we need to set IgnoreClusterId to true, when we already get the version from getDefaultProtocolVersionMsg() so we already know that the old version of source does not have ClusterId?

Reply:
We actually don't use either of the fields set here...Its legacy code now, and my guess for the setting of it is that since LR V1 reuses some of the constructs of corfu, and corfu set the 2 fields that way, LR ended up doing the same. 
I have copied the request exactly as how the request from LR V1 would look like.